packages/checkpoint-dev/checkpoint.sh

```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+
+# Name of the running k3d container
+K3S_CONTAINER="k3d-uds-server-0"
+
+if [ -z "$TMPDIR" ]; then
+    #  macOS sets TMPDIR to a user temp directory - this also provides more options to linux
+    TMPDIR="/tmp"
+fi
+DATA_DIR="${TMPDIR}/uds-checkpoint-data"
+
+# Step 0: Ensure we can get sudo
+echo "This package requires elevated permissions to create - requesting sudo (if paused enter password)"
+sudo echo "got sudo! success!"
+
+# Step 1: Get the container ID of the running k3d container
+CONTAINER_ID=$(docker ps -qf "name=$K3S_CONTAINER")
+
+if [ -z "$CONTAINER_ID" ]; then
+    echo "No running container found for $K3S_CONTAINER"
+    exit 1
+fi
+
+# Step 2: Get the mounted volumes of the running container
+echo "Inspecting container volumes for $CONTAINER_ID..."
+VOLUMES=$(docker inspect -f '{{ json .Mounts }}' "$CONTAINER_ID" | jq)
+
+# Step 3: Prepare directories to save the volume data
+sudo rm -rf "$DATA_DIR"
+mkdir -p "${DATA_DIR}/kubelet_data" "${DATA_DIR}/k3s_data"
+
+# Step 4: Loop through volumes and copy data to corresponding directories
+echo "Copying volumes to local directories..."
+
+for row in $(echo "$VOLUMES" | jq -r '.[] | @base64'); do
+    _jq() {
+        echo "${row}" | base64 --decode | jq -r "${1}"
+    }
+
+    SOURCE=$(_jq '.Source')
+    DESTINATION=$(_jq '.Destination')
+
+    case "$DESTINATION" in
+        "/var/lib/kubelet")
+            echo "Copying $SOURCE to ${DATA_DIR}/kubelet_data/"
+            sudo cp -a "$SOURCE"/. "${DATA_DIR}/kubelet_data/"
+            ;;
+        "/var/lib/rancher/k3s")
+            echo "Copying $SOURCE to ${DATA_DIR}/k3s_data/"
+            sudo cp -a "$SOURCE"/. "${DATA_DIR}/k3s_data/"
+            ;;
```

Comment:
During creation I see these errors (which cause the deploy to fail later):
```
     cp: /var/lib/docker/volumes/c0d8ea4ead46f3c6649218be409e19d1cd63bfcc68f32d548a116c7924d7a793/_data/.: No such file or directory
     cp: /var/lib/docker/volumes/822e843b8cf644f9c4c9118671f6014d32ad84a062d690e69b07d5c6fdfcfbe2/_data/.: No such file or directory
```

I think pretty much universally on macOS docker is run inside of a VM, in my case the VM can be accessed with `colima ssh` but docker desktop, rancher desktop, etc would likely have similar issues and ways to access the VM.

I was able to rewrite a portion of this script to use [`docker cp`](https://docs.docker.com/reference/cli/docker/container/cp/) instead and got closer (at least didn't get errors with the volumes). I think this is probably a better, more agnostic option here and simplifies a lot of this logic (no looping through volumes, just copy the two paths we need explicitly). I was hoping it might also remove the need for `sudo` but in my case one of the paths gave some permission errors still until I added sudo. I'm sure there's some efficiency loss here, but since it's create time I think it's worth it to make this work across distros? In my run locally it took less than a minute still to run which still seems decently performant (granted I couldn't get it to run successfully previously so unsure of the real comparison).

Would be curious your thoughts on this - I dropped the script changes into a gist since there were a handful of changes across the entirety of the file: https://gist.github.com/mjnagel/6d681678df83067169c4e652466f704f

I also had to add `--no-xattrs` to the final tar command, I got warnings/errors without this (suspect that's some macOS <> Linux stuff). This got me much closer but I hit some issues with the token:
```
time="2024-10-02T15:19:18Z" level=fatal msg="starting kubernetes: preparing server: bootstrap data already found and encrypted with different token"
```

I tried to tweak the commands around startup (using the k3d `--token` option rather than k3s arg) and validated the token exists after extraction but couldn't figure this one out. Would be curious if you hit the same issue with my modified script and can figure out what's wrong?

Reply:
per discussion ignoring macOS for now - we can revisit at a later time - but would like to use this to speed up CI while we think about macOS support